### ISSUE TITLE
Fail in configuring modules with exotic MB and EUMM setups

### ISSUE THREAD

**Issue by zakame** [skaji/cpm#74]
Hi @skaji, firstly thanks for making App::cpm!

I'm trying to install for a local dist having this cpanfile:

```perl 
requires 'Mojolicious', '7.0';
requires 'Data::Utilities', '0.04';
```

However, `cpm` fails to do so:

```sh
θ61° [zakame:~/hoozuki] 21s % cpm install
DONE install Module-Build-0.4220
DONE install Mojolicious-7.29
FAIL install Data-Utilities-0.04
2 distributions installed.  
See /home/zakame/.perl-cpm/build.log for details. 
```

Checking the log shows this:

```
2017-03-22T14:16:36,1829,Data::Utilities| Resolved Data::Utilities (0.04) -> http://www.cpan.org/authors/id/C/CO/CORNELIS/Data-Utilities-0.04.tar.gz from MetaDB
2017-03-22T14:16:36,1829,Data-Utilities-0.04| Using cache /home/zakame/.perl-cpm/cache/authors/id/C/CO/CORNELIS/Data-Utilities-0.04.tar.gz
2017-03-22T14:16:36,1829,Data-Utilities-0.04| Unpacking Data-Utilities-0.04.tar.gz
2017-03-22T14:16:36,1829,Data-Utilities-0.04| Configuring distribution
2017-03-22T14:16:36,1851,Data-Utilities-0.04| Checking if your kit is complete... 
2017-03-22T14:16:36,1851,Data-Utilities-0.04| Looks good
2017-03-22T14:16:36,1851,Data-Utilities-0.04| Generating a Unix-style Makefile
2017-03-22T14:16:36,1851,Data-Utilities-0.04| Writing Makefile for Data::Utilities
2017-03-22T14:16:36,1851,Data-Utilities-0.04| Writing MYMETA.yml and MYMETA.json
2017-03-22T14:16:36,1829,Data-Utilities-0.04| ! Retrying (you can turn off this behavior by --no-retry) 2017-03-22T14:16:36,1922,Data-Utilities-0.04| Generating a Unix-style Makefile
2017-03-22T14:16:36,1922,Data-Utilities-0.04| Writing Makefile for Data::Utilities
2017-03-22T14:16:36,1922,Data-Utilities-0.04| Writing MYMETA.yml and MYMETA.json
2017-03-22T14:16:36,1829,Data-Utilities-0.04| Failed to configure distribution
```

Taking a look with `cpanm --look Data::Utilities` show a quite unusual setup of having both a `Build.PL` and `Makefile.PL`, (since it is an older module that uses `Module::Install`,) and furthermore, its `Build.PL` produces a `Makefile` instead of the expected `Build`:

```
θ62° [zakame:~] 9s % cpanm --look Data::Utilities 
--> Working on Data::Utilities Fetching file:///home/ftp/pub/CPAN/authors/id/C/CO/CORNELIS/Data-Utilities-0.04.tar.gz ... OK
Entering /home/zakame/.cpanm/work/1490163513.5051/Data-Utilities-0.04 with /bin/zsh
θ68° [zakame:~/.cpanm … Data-Utilities-0.04] % ls 
Build.PL lib Makefile.PL META.yml t Changes LICENSE MANIFEST README Todo 
θ61° [zakame:~/.cpanm … Data-Utilities-0.04] % cat Build.PL 
# Dear Distribution Packager. This use of require is intentional.  
# Module::Install detects Build.PL usage and acts accordingly.  
require 'Makefile.PL'; 
θ63° [zakame:~/.cpanm … Data-Utilities-0.04] % perl Build.PL 
Checking if your kit is complete...  Looks good 
Generating a Unix-style Makefile 
Writing Makefile for Data::Utilities 
Writing MYMETA.yml and MYMETA.json 
θ63° [zakame:~/.cpanm … Data-Utilities-0.04] % ls 
Build.PL lib Makefile MANIFEST MYMETA.json README Todo 
Changes LICENSE Makefile.PL META.yml MYMETA.yml t
```

Digging around `App::cpm::Worker::Installer->configure()`, I found that it doesn't try different configurators until it can get a proper `Build` or `Makefile`.  I'll make a PR that tries to fix this.

**Comment by skaji** [comment]
Thanks for reporting this.

I will accept your pull request, but let me ask you a question first:

*Do you really think CPAN clients should support the installation of this type of CPAN modules?*

Actually I don't think so. Once a CPAN client support it, then people would say that:

* CPAN client **A** support **X**, so CPAN client **B** must support **X** too
* CPAN client **A** supoort **X**, so it must support **Y** too

I think it is important to fix the spec CPAN modules, and to exclude modules which does not follow the spec. Then CPAN ecosystem will be simpler, and people can hack CPAN easily.

Note: a draft spec of Build.PL is here https://github.com/Perl-Toolchain-Gang/cpan-api-buildpl/blob/master/lib/CPAN/API/BuildPL.pm

**Comment by zakame** [comment]
Thanks for considering my PR! :+1: 

To answer the question,

>Do you really think CPAN clients should support the installation of this type of CPAN modules?

In my case, I'm someone supporting current systems using `carton`, so I was exploring moving from to `cpm` in the hopes to speed up deployment times.  My particular codebase is using `Data::Utilities` along with some 400 other CPAN modules (most using standard builds using either MB or EUMM only, and some others rolling their own,) installed nicely by `carton` in local and working well.  So it was to my great surprise that `cpm` could not install it...

...and now, you have this issue and PR :grinning: 

In short, I think CPAN clients should strive to be able to do their job of fetching, configuring, building, testing, and installing CPAN modules to the best they can, within reason.  I think it is not unreasonable for e.g. a module having Build.PL and Makefile.PL (like in the `Data::Utilities` case) and CPAN clients should be able to use either; on the other hand, it may be unreasonable to support even more exotic setups (such as a module using cmake or easy_install instead of MB/EUMM.)
